EIPS/eip-6404.md

```diff
@@ -0,0 +1,1000 @@
+    value: uint256
+    data: ByteList[MAX_CALLDATA_SIZE]
+    access_list: List[AccessTuple, MAX_ACCESS_LIST_SIZE]  # EIP-2930
+    blob_versioned_hashes: List[VersionedHash, MAX_VERSIONED_HASHES_LIST_SIZE]  # EIP-4844
+
+class ECDSASignature(Container):
+    y_parity: boolean  # EIP-2930
+    r: uint256
+    s: uint256
+
+class SignedTransaction(Container):
+    tx: Transaction
+    signature: ECDSASignature
+    hash_version: TxHashVersion
+
+class BlobTransactionExtension(Container):
+    blob_kzgs: List[KZGCommitment, MAX_TX_WRAP_KZG_COMMITMENTS]
+    blobs: List[Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB], LIMIT_BLOBS_PER_TX]
+    kzg_aggregated_proof: KZGProof
+
+class TransactionPayload(Container):
+    signed_tx: SignedTransaction
+    blob: Optional[BlobTransactionExtension]  # EIP-4844
+```
+
+On the network, `TransactionPayload` is preceded by a single byte, [`BLOB_TX_TYPE`](./eip-4844.md#parameters) to distinguish from other transaction formats according to [EIP-2718](./eip-2718.md), regardless of the inner `signed_tx.hash_version`.
+
+The `blocks.transactions` list remains RLP encoded:
+
+```python
+block.transactions = RLP([
+    SSZ.encode(transaction_0),
+    SSZ.encode(transaction_1),
+    SSZ.encode(transaction_2),
+    ...
+])
+```
+
+#### Consensus `ExecutionPayload` changes (Transactions)
+
+The [CL `ExecutionPayload`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/eip4844/beacon-chain.md#executionpayload) is updated to replace the existing [CL `Transaction`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/bellatrix/beacon-chain.md#custom-types) container with the new `SignedTransaction` container.
+
+| Name | Value | Description |
+| - | - | - |
+| [`MAX_TRANSACTIONS_PER_PAYLOAD`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/bellatrix/beacon-chain.md#execution) | `uint64(2**20)` (= 1,048,576) | Maximum amount of transactions allowed in each block |
+
+```python
+class ExecutionPayload(Container):
+    ...
+    transactions: List[SignedTransaction, MAX_TRANSACTIONS_PER_PAYLOAD]  # Formerly, `Transaction`
+    ...
+```
+
+#### Execution block header changes (Transactions)
+
+The execution block header's `transactions_root` is updated to the `transactions_root` definition of the [consensus-specs `ExecutionPayloadHeader`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/eip4844/beacon-chain.md#executionpayloadheader).
+
+To compute the new `transactions_root`, the list of individual `SignedTransaction` containers is represented as a SSZ `List`.
+
+```python
+block_header.transactions_root == hash_tree_root(List[SignedTransaction, MAX_TRANSACTIONS_PER_PAYLOAD](
+    transaction_0,
+    transaction_1,
+    transaction_2,
+    ...
+)) == hash_tree_root(List[Root, MAX_TRANSACTIONS_PER_PAYLOAD](
```

Comment:
isn't this equality a given based on the logic of HTR? it seems unnecessary to explicitly call out the recursive merkleization

Reply:
Yes, this is just a property of HTR. Removing from the spec section.